[DashboardLayout] Support route patterns in navigation (#3991)
mui/mui-toolpad: docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md

### Navigation Actions

Navigation links have an optional `action` prop to render any content on the right-side of the respective list item, such as badges with numbers, or buttons to toggle a popover menu.

{{"demo": "DashboardLayoutNavigationActions.js", "height": 400, "iframe": true}}

### Navigation Pattern Matching

Navigation links have an optional `pattern` prop to define a pattern to be matched for the item to be marked as selected.

> Any path supported by the [path-to-regexp](https://www.npmjs.com/package/path-to-regexp) library can be used, such as:

Reviewer: I wouldn't encourage users too much 😄 
```suggestion
This feature is built on top of the [path-to-regexp](https://www.npmjs.com/package/path-to-regexp) library.
```

- Standard paths: `/foo`
- Named parameters: `/:foo/:bar`

> - Unnamed parameters: `/:foo/(.*)`

Reviewer: I don't think named vs unnamed is relevant to us. I'd just say "Parameters" on the line above. I'd also use something slightly more real-life as an example, like "orders"

 ```md
 - Constant path: `/orders`
 - Dynamic segment: `/orders/:segment`
 - Zero or more segments: `/orders{/:segment}*`
 - One or more segments: `/orders{/:segment}+`
 - Optional segment: `/orders{/:segment}+?`
```

> - Optional parameters: `/:foo{/:bar}?`

Reviewer: We could maybe add demo. a "product" page that has links to show the "next product" and the current path. e.g.

```
current path: /products/20
title: Item 20
[next product](/products/21)
```
